apex/contrib/csrc/nccl_allocator/NCCLAllocator.cpp

```diff
@@ -0,0 +1,437 @@
+           md != nccl_allocation_metadata_.end();
+	   md++
+    ) {
+      // compare stream?
+      if(md->second.is_free && md->second.size == size && md->second.device_idx == device) {
+        md->second.is_free = false;
+	r = md->second.ptr;
+	return r;
+      }
+    }
+
+    C10_NCCL_CHECK(ncclMemAlloc(&r, size));
+    {
+      const std::lock_guard<std::mutex> lock(allocator_mutex_);
+      nccl_allocation_metadata_.emplace(r, _AllocationMetadata(size, device, stream, r));
+
+      auto te = c10::cuda::CUDACachingAllocator::TraceEntry(
+        c10::cuda::CUDACachingAllocator::TraceEntry::Action::SEGMENT_ALLOC,
+        device,
+        (int64_t)r,
+        size,
+        stream,
+        c10::getApproximateTime(),
+        nullptr);
+      for (const auto& cb : trace_trackers_) {
+        cb(te);
+      }
+    }
+  }
+#else
+  if (_use_nccl_mem) {
+    TORCH_WARN_ONCE("ncclMemAlloc is not supported. ",
+                    "NCCL v2.19 and newer is required,",
+                    "but found NCCL v", NCCL_MAJOR, ".", NCCL_MINOR, ". ",
+                    "Falling back to cudaMalloc.");
+  }
+  if (false) {
+    // goto cudaMalloc
+  }
+#endif
+  else {
+    if (r == nullptr) {
+      // find allocation of the same size or slightly better
+      // What is the best upper bound?
+      auto start = cuda_rt_cache_metadata_.lower_bound(size);
+      auto end = cuda_rt_cache_metadata_.upper_bound(size + 1024);
+      for (auto md = start; md != end; ) {
+        _AllocationMetadata metadata = md->second;
+        // Should we check for stream?
+        if (metadata.device_idx == device) {
+          r = metadata.ptr;
+          cuda_rt_cache_metadata_.erase(md++);
+          break;
+        }
+        ++md;
+      }
+    }
+    if (r == nullptr) {
+      cudaError_t err = cudaMalloc(&r, size);
+      if (cudaSuccess != err) {
+        size_t freed = 0;
+        for (auto md = cuda_rt_cache_metadata_.rbegin(); md != cuda_rt_cache_metadata_.rend(); ++md) {
+          _AllocationMetadata metadata = md->second;
+          freed = freed + metadata.size;
+          cuda_rt_cache_metadata_.erase(--(md.base()));
+	  C10_CUDA_CHECK(cudaFree(metadata.ptr));
+          if (freed >= size) {
+            break;
+          }
+        }
+        C10_CUDA_CHECK(cudaMalloc(&r, size));
+      }
+    }
+    {
+      const std::lock_guard<std::mutex> lock(allocator_mutex_);
+      cuda_rt_allocation_metadata_.emplace(r, _AllocationMetadata(size, device, stream, r));
+    }
+  }
+  return r;
+}
+
+c10::DataPtr NCCLAllocator::allocate(size_t size) {
+  c10::DeviceIndex device = -1;
+  C10_CUDA_CHECK(c10::cuda::GetDevice(&device));
+  cudaStream_t stream =
+      c10::cuda::getCurrentCUDAStream(static_cast<int>(device));
+  void* r = this->malloc(size, device, stream);
+  c10::DataPtr data_ptr = {
+      r,
+      r,
+      raw_deleter(),
+      c10::Device(
+          c10::DeviceType::CUDA, static_cast<int>(device))};
```

Comment:
it doesn't seem like this cast is required -- https://github.com/pytorch/pytorch/blob/768ce2cddad2057349d1194274a5f93c47c5ac88/c10/core/Device.h#L34-L39